staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1864,3 +1865,126 @@ func TestForgetWatcher(t *testing.T) {
 	assertCacherInternalState(0, 0)
 	require.Equal(t, 2, forgetCounter)
 }
+
+func TestStreamSeparation(t *testing.T) {
+	tcs := []struct {
+		name                         string
+		separateCacheWatchRPC        bool
+		useWatchCacheContextMetadata bool
+		expectBookmarkOnWatchCache   bool
+		expectBookmarkOnEtcd         bool
+	}{
+		{
+			name:                       "common RPC > both get bookmarks",
+			separateCacheWatchRPC:      false,
+			expectBookmarkOnEtcd:       true,
+			expectBookmarkOnWatchCache: true,
+		},
+		{
+			name:                         "common RPC & watch cache context > both get bookmarks",
+			separateCacheWatchRPC:        false,
+			useWatchCacheContextMetadata: true,
+			expectBookmarkOnEtcd:         true,
+			expectBookmarkOnWatchCache:   true,
+		},
+		{
+			name:                       "separate RPC > only etcd gets bookmarks",
+			separateCacheWatchRPC:      true,
+			expectBookmarkOnEtcd:       true,
+			expectBookmarkOnWatchCache: false,
+		},
+		{
+			name:                         "separate RPC & watch cache context > only watch cache gets bookmarks",
+			separateCacheWatchRPC:        true,
+			useWatchCacheContextMetadata: true,
+			expectBookmarkOnEtcd:         false,
+			expectBookmarkOnWatchCache:   true,
+		},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SeparateCacheWatchRPC, tc.separateCacheWatchRPC)()
+			server, etcdStorage := newEtcdTestStorage(t, etcd3testing.PathPrefix())
+			_, cacher, server, terminate := testSetupCacher(t, server, etcdStorage)
+			if err := cacher.ready.wait(context.Background()); err != nil {
+				t.Fatalf("unexpected error waiting for the cache to be ready")
+			}
+			t.Cleanup(terminate)
+
+			getCacherRV := func() uint64 {
+				cacher.watchCache.RLock()
+				defer cacher.watchCache.RUnlock()
+				return cacher.watchCache.resourceVersion
+			}
+			waitContext, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			defer cancel()
+			waitForEtcdBookmarkRevision := watchAndWaitForBookmarkRevision(waitContext, etcdStorage)
+
+			rv := getCacherRV()
+			resp, err := server.V3Client.Put(context.Background(), "foo", "bar")
```

Comment:
Why can't you use etcdStorage? etcdStorage doesn't assume anything about keys and you can use arbitrary key, so `etcdStorage.Create(ctx, "foo", &v1.Pod{}, out, 0)` should do the job.

Reply:
Sorry, didn't get to fixing that yet. Got stuck on ` panic: no kind is registered for the type v1.Pod in scheme "k8s.io/apiserver/pkg/storage/cacher/cacher_testing_utils_test.go:43"`.

Reply:
ahh sorry - you should use examplev1.Pod instead